File: macos/WeatherOddsCore/Sources/WeatherOddsCore/Retry.swift
import Foundation

/// HTTP-level detail from a failed upstream response.
///
/// This is captured before the body is decoded so a rate-limit response with a
/// non-JSON body still carries its status code and retry headers. The
/// `Retry-After` value is kept verbatim: resolving its HTTP-date form needs the
/// reference date used by the retry policy, not the moment the header arrived.
public struct UpstreamHTTPFailure: Codable, Equatable, Sendable {
    public let statusCode: Int
    public let retryAfterHeader: String?
    public let reason: String?

    public init(statusCode: Int, retryAfterHeader: String? = nil, reason: String? = nil) {
        self.statusCode = statusCode
        self.retryAfterHeader = retryAfterHeader
        self.reason = reason
    }

    /// Whether the upstream asked every caller to stop, not just this request.
    public var isRateLimited: Bool {
        statusCode == 429
    }
}

/// `Retry-After` parsing for both forms allowed by RFC 9110.
public enum RetryAfterHeader {
    /// Returns the delay the server asked for, or nil when it is absent,
    /// malformed, or already elapsed. A non-positive delay is treated as absent
    /// so the caller falls back to its own backoff instead of retrying at once.
    public static func delay(_ value: String?, at now: Date) -> TimeInterval? {
        guard let trimmed = value?.trimmingCharacters(in: .whitespacesAndNewlines),
              !trimmed.isEmpty
        else {
            return nil
        }

        // delta-seconds is an unsigned integer; anything else is an HTTP-date.
        if trimmed.allSatisfy(\.isNumber), let seconds = Double(trimmed) {
            return seconds > 0 ? seconds : nil
        }

        guard let date = httpDate(trimmed) else { return nil }
        let delay = date.timeIntervalSince(now)
        return delay > 0 ? delay : nil
    }

    private static func httpDate(_ value: String) -> Date? {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .gmt
        for format in [
            "EEE, dd MMM yyyy HH:mm:ss zzz",   // IMF-fixdate
            "EEEE, dd-MMM-yy HH:mm:ss zzz",    // obsolete RFC 850
            "EEE MMM d HH:mm:ss yyyy",         // obsolete asctime
        ] {
            formatter.dateFormat = format
            if let date = formatter.date(from: value) { return date }
        }
        return nil
    }
}

/// Decides when a failed refresh may be attempted again.
///
/// Retries are always handed to WidgetKit as a timeline reload date; the
/// extension never sleeps or loops waiting for a deadline.
public struct RetryPolicy: Sendable {
    /// Delay after a single transient failure.
    public static let baseDelay: TimeInterval = 30 * 60
    /// Documented ceiling for locally computed backoff, matching the interval
    /// between successful refreshes.
    public static let maximumDelay: TimeInterval = 6 * 60 * 60
    /// Ceiling applied to a server-supplied `Retry-After`, which may exceed
    /// `maximumDelay` but must not park the widget indefinitely.
    public static let maximumRetryAfter: TimeInterval = 24 * 60 * 60
    /// Fraction of the computed delay that jitter may subtract. Subtracting
    /// keeps successive delays ordered and never exceeds the cap.
    public static let jitterFraction = 0.2
    /// Retry delay after a cancelled refresh, which records no failure state.
    public static let interruptedDelay: TimeInterval = 5 * 60

    /// Returns a value in `0..<1`. Injectable so tests are deterministic.
    private let jitter: @Sendable () -> Double

    public init(jitter: @escaping @Sendable () -> Double = { Double.random(in: 0..<1) }) {
        self.jitter = jitter
    }

    /// Capped exponential backoff for the nth consecutive failure, where the
    /// first failure is 1.
    public func delay(forConsecutiveFailures failures: Int) -> TimeInterval {
        // Clamping the exponent keeps a corrupt or absurd counter finite.
        let exponent = min(max(failures, 1) - 1, 16)
        let capped = min(Self.baseDelay * pow(2, Double(exponent)), Self.maximumDelay)
        let fraction = min(max(jitter(), 0), 1)
        return capped * (1 - Self.jitterFraction * fraction)
    }

    /// The earliest permitted next attempt. A usable `Retry-After` wins over
    /// local backoff, including when it is longer than `maximumDelay`.
    public func nextAttempt(
        at now: Date,
        consecutiveFailures: Int,
        httpFailure: UpstreamHTTPFailure? = nil
    ) -> Date {
        if let requested = RetryAfterHeader.delay(httpFailure?.retryAfterHeader, at: now) {
            return now.addingTimeInterval(min(requested, Self.maximumRetryAfter))
        }
        return now.addingTimeInterval(delay(forConsecutiveFailures: consecutiveFailures))
    }
}

/// Durable cooldown state for one retry key.
///
/// `recordedAt` is retained so a clock change cannot leave a widget parked: a
/// record written in the future is ignored, and a deadline further out than
/// `RetryPolicy.maximumRetryAfter` is clamped.
public struct RetryState: Codable, Equatable, Sendable {
    public static let currentSchemaVersion = 1

    public let schemaVersion: Int
    public let key: String
    public let consecutiveFailures: Int
    public let recordedAt: Date
    public let nextAttemptAfter: Date

    public init(
        schemaVersion: Int = RetryState.currentSchemaVersion,
        key: String,
        consecutiveFailures: Int,
        recordedAt: Date,
        nextAttemptAfter: Date
    ) {
        self.schemaVersion = schemaVersion
        self.key = key
        self.consecutiveFailures = consecutiveFailures
        self.recordedAt = recordedAt
        self.nextAttemptAfter = nextAttemptAfter
    }

    /// The deadline this record imposes at `now`, or nil when an attempt is
    /// permitted.
    public func activeDeadline(at now: Date) -> Date? {
        guard now >= recordedAt else { return nil }
        let deadline = min(
            nextAttemptAfter,
            recordedAt.addingTimeInterval(RetryPolicy.maximumRetryAfter)
        )
        return now < deadline ? deadline : nil
    }

    /// The failure count a following failure should build on. A record whose
    /// deadline has long passed still counts, so a persistently broken upstream
    /// does not restart at the base delay on every wake-up.
    public func failureCount(at now: Date) -> Int {
        now >= recordedAt ? consecutiveFailures : 0
    }
}

/// Why a refresh produced no forecast, and when the next attempt is permitted.
public struct ForecastUnavailable: Error, Equatable, Sendable {
    public enum Reason: Equatable, Sendable {
        /// Suppressed by a durable cooldown before any request was made.
        case cooldown
        /// The upstream returned HTTP 429 for this or another configuration.
        case rateLimited
        /// Any other retryable failure, including transport errors.
        case transient
    }

    public let reason: Reason
    public let nextAttempt: Date
    public let consecutiveFailures: Int
    public let httpFailure: UpstreamHTTPFailure?
    public let message: String?

    public init(
        reason: Reason,
        nextAttempt: Date,
        consecutiveFailures: Int,
        httpFailure: UpstreamHTTPFailure? = nil,
        message: String? = nil
    ) {
        self.reason = reason
        self.nextAttempt = nextAttempt
        self.consecutiveFailures = consecutiveFailures
        self.httpFailure = httpFailure
        self.message = message
    }
}

extension ForecastUnavailable: LocalizedError {
    public var errorDescription: String? {
        switch reason {
        case .cooldown:
            "Forecast refresh is waiting for its next permitted attempt."
        case .rateLimited:
            "The forecast service asked callers to slow down."
        case .transient:
            message ?? "Forecast service is temporarily unavailable."
        }
    }
}

File: src/weatherodds/retry.py
"""Retry budget, ``Retry-After`` parsing, and durable cooldown records.

The CLI is interactive, so the inline retry budget is deliberately small: a
couple of quick jittered attempts measured in seconds. A delay the server asks
for that is longer than :attr:`RetryPolicy.max_inline_wait` is never slept on;
it becomes a cooldown persisted next to the forecast cache, so the next
invocation reports the deadline instead of hammering an upstream that already
said no.
"""

from __future__ import annotations

import datetime as dt
import email.utils
from dataclasses import dataclass

# Ceiling applied to any persisted cooldown, including one the server asked
# for. It matches the forecast freshness window: while a cooldown is in force a
# cached forecast from within that window is still served, so the cap bounds
# how long a stale clock or an absurd header can keep the CLI offline.
MAX_COOLDOWN = 6 * 60 * 60

# Clamps the exponent so a corrupt or absurd failure counter stays finite.
MAX_BACKOFF_EXPONENT = 16


def _clamp_unit(value: float) -> float:
    return min(max(value, 0.0), 1.0)


@dataclass(frozen=True)
class UpstreamFailure:
    """HTTP-level detail from a failed upstream response.

    Captured before the body is decoded, so a rate-limit response with an HTML
    body still carries its status code and headers. ``retry_after`` is kept
    verbatim: resolving its HTTP-date form needs the clock reading used by the
    retry policy, not the moment the header arrived.
    """

    status_code: int
    retry_after: str | None = None
    reason: str | None = None

    @property
    def is_rate_limited(self) -> bool:
        """Whether the upstream asked every caller to stop, not just this one."""
        return self.status_code == 429

    def retry_after_delay(self, now: float) -> float | None:
        return parse_retry_after(self.retry_after, now)


def parse_retry_after(value: str | None, now: float) -> float | None:
    """Seconds to wait per RFC 9110 ``Retry-After``, or None when unusable.

    Both forms are accepted: delta-seconds and an HTTP-date. A missing,
    malformed, or already-elapsed value returns None so the caller falls back
    to its own backoff rather than retrying immediately.
    """
    if value is None:
        return None
    text = value.strip()
    if not text:
        return None

    # delta-seconds is an unsigned integer; anything else is an HTTP-date.
    if text.isdigit():
        seconds = float(text)
        return seconds if seconds > 0 else None

    try:
        parsed = email.utils.parsedate_to_datetime(text)
    except (TypeError, ValueError):
        return None
    if parsed is None:
        return None
    if parsed.tzinfo is None:
        parsed = parsed.replace(tzinfo=dt.timezone.utc)
    delay = parsed.timestamp() - now
    return delay if delay > 0 else None


@dataclass(frozen=True)
class RetryPolicy:
    """The documented retry budget for one CLI invocation.

    Inline retries cover transport errors and 5xx responses: at most
    ``max_attempts`` requests per URL, separated by capped exponential backoff
    with subtractive jitter, and never running past ``max_elapsed`` seconds of
    waiting. Rate limiting is not retried inline at all — it is recorded as a
    cooldown, which starts at ``cooldown_base`` and doubles per consecutive
    rate-limited invocation whenever the server sends no usable
    ``Retry-After``.
    """

    base_delay: float = 0.5
    multiplier: float = 2.0
    max_delay: float = 8.0
    max_attempts: int = 3
    max_elapsed: float = 20.0
    # Jitter is subtracted, so successive delays stay ordered and never exceed
    # the cap.
    jitter_fraction: float = 0.2
    max_inline_wait: float = 5.0
    cooldown_base: float = 60.0
    max_cooldown: float = MAX_COOLDOWN

    def backoff(self, attempt: int, jitter: float = 0.0) -> float:
        """Delay before retrying, where the first completed attempt is 1."""
        exponent = min(max(attempt, 1) - 1, MAX_BACKOFF_EXPONENT)
        capped = min(self.base_delay * self.multiplier**exponent, self.max_delay)
        return capped * (1 - self.jitter_fraction * _clamp_unit(jitter))

    def cooldown_delay(
        self, consecutive_failures: int, retry_after: float | None = None
    ) -> float:
        """How long to stay off Open-Meteo after a rate-limited response.

        A usable ``Retry-After`` wins over local backoff, still capped by
        ``max_cooldown``.
        """
        if retry_after is not None:
            return min(retry_after, self.max_cooldown)
        exponent = min(max(consecutive_failures, 1) - 1, MAX_BACKOFF_EXPONENT)
        return min(self.cooldown_base * self.multiplier**exponent, self.max_cooldown)


@dataclass(frozen=True)
class RetryState:
    """Durable cooldown for one retry key.

    ``recorded_at`` is kept so a clock change cannot park the CLI: a record
    written in the future is ignored, and a deadline further out than
    :data:`MAX_COOLDOWN` past its recording is clamped.
    """

    SCHEMA_VERSION = 1

    key: str
    consecutive_failures: int
    recorded_at: float
    next_attempt_after: float

    def active_deadline(self, now: float, max_cooldown: float = MAX_COOLDOWN) -> float | None:
        """The deadline this record imposes now, or None when a call is allowed."""
        if now < self.recorded_at:
            return None
        deadline = min(self.next_attempt_after, self.recorded_at + max_cooldown)
        return deadline if now < deadline else None

    def failure_count(self, now: float) -> int:
        """The count a following failure builds on.

        An expired record still counts, so a persistently rate-limited upstream
        does not restart at the base delay on every invocation.
        """
        return self.consecutive_failures if now >= self.recorded_at else 0

    def to_dict(self) -> dict:
        return {
            "schema_version": self.SCHEMA_VERSION,
            "key": self.key,
            "consecutive_failures": self.consecutive_failures,
            "recorded_at": self.recorded_at,
            "next_attempt_after": self.next_attempt_after,
        }

    @classmethod
    def from_dict(cls, data: object, key: str) -> RetryState | None:
        """Rebuild a record, or None when it is corrupt or for another key."""
        if not isinstance(data, dict) or data.get("schema_version") != cls.SCHEMA_VERSION:
            return None
        if data.get("key") != key:
            return None
        try:
            failures = int(data["consecutive_failures"])
            recorded_at = float(data["recorded_at"])
            next_attempt_after = float(data["next_attempt_after"])
        except (KeyError, TypeError, ValueError):
            return None
        if failures < 1:
            return None
        return cls(
            key=key,
            consecutive_failures=failures,
            recorded_at=recorded_at,
            next_attempt_after=next_attempt_after,
        )
